### PR TITLE
INTYGFV-16023: Implemented service to handle request to intyg proxy service for getCareProviderOfCareUnit

### PR DESCRIPTION
**Pull request by mikaelaromild1**

### CODE REVIEW

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/client/HsaIntygProxyServiceCareProviderOfCareUnitClient.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+import se.inera.intyg.infra.integration.hsatk.exception.HsaServiceCallException;
+import se.inera.intyg.infra.integration.intygproxyservice.dto.GetCareProviderOfCareUnitRequestDTO;
+import se.inera.intyg.infra.integration.intygproxyservice.dto.GetCareProviderOfCareUnitResponseDTO;
+
+@Service
+public class HsaIntygProxyServiceCareProviderOfCareUnitClient {
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/client/HsaIntygProxyServiceCareProviderOfCareUnitClient.java:32]: This client will be reused any time we want to get a HealthCareUnit. Change the name to HsaIntygProxyServiceHealthCareUnitClient

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/client/HsaIntygProxyServiceCareProviderOfCareUnitClient.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Autowired
+    @Qualifier("hsaIntygProxyServiceRestTemplate")
+    private RestTemplate restTemplate;
+    @Value("${integration.intygproxyservice.healthcareunit.endpoint}")
+    private String careProviderOfCareUnitEndpoint;
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/client/HsaIntygProxyServiceCareProviderOfCareUnitClient.java:38]: Rename to healthCareUnitEndpoint

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/client/HsaIntygProxyServiceCareProviderOfCareUnitClient.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    @Value("${integration.intygproxyservice.baseurl}")
+    private String intygProxyServiceBaseUrl;
+
+    public GetCareProviderOfCareUnitResponseDTO getCareProviderOfCareUnit(
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/client/HsaIntygProxyServiceCareProviderOfCareUnitClient.java:42]: Rename method to getHealthCareUnit(...)

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/dto/GetCareProviderOfCareUnitResponseDTO.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import se.inera.intyg.infra.integration.hsatk.model.HealthCareUnit;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class GetCareProviderOfCareUnitResponseDTO {
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/dto/GetCareProviderOfCareUnitResponseDTO.java:32]: Rename to HealthCareUnitResponseDTO

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/dto/GetCareProviderOfCareUnitRequestDTO.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package se.inera.intyg.infra.integration.intygproxyservice.dto;
+
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+public class GetCareProviderOfCareUnitRequestDTO {
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/dto/GetCareProviderOfCareUnitRequestDTO.java:27]: Rename to HealthCareUnitRequestDTO

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/constants/HsaIntygProxyServiceConstants.java [hunk 6]
```diff
@@ -26,7 +26,8 @@ private HsaIntygProxyServiceConstants() {
     }
 
     public static final String EMPLOYEE_CACHE_NAME = "hsaIntygProxyServiceEmployeeCache";
-    public static final String HEALTH_CARE_UNIT_MEMBERS_CACHE_NAME = "hsaIntygProxyServiceHealthCareUnitMembersCache";
+    public static final String CARE_PROVIDER_OF_CARE_UNIT_CACHE_NAME = "hsaIntygProxyServiceCareProviderOfCareUnitCache";
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/constants/HsaIntygProxyServiceConstants.java:29]: Rename to HEALTH_CARE_UNIT_CACHE_NAME = "hsaIntygProxyServiceHealthCareUnitCache"

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/dto/GetCareProviderOfCareUnitRequestDTO.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    String careUnitHsaId;
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/dto/GetCareProviderOfCareUnitRequestDTO.java:29]: Rename member to hsaId so it matches the api in intyg-proxy-service

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/services/HsaLegacyIntegrationOrganizationService.java [hunk 8]
```diff
@@ -49,7 +51,17 @@ public UserAuthorizationInfo getAuthorizedEnheterForHosPerson(String hosPersonHs
...
-        return null;
+        try {
+            return getCareProviderOfCareUnitService.get(
+                GetCareProviderOfCareUnitRequestDTO.builder()
+                    .careUnitHsaId(vardenhetHsaId)
+                    .build()
+            );
+
+        } catch (HsaServiceCallException hsaServiceCallException) {
+            log.warn("HsaServiceCallException thrown: {}", hsaServiceCallException);
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/services/HsaLegacyIntegrationOrganizationService.java:62]: Consider changing the log message to something more descriptive e.g. "Could fetch health care unit".

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/services/GetCareProviderOfCareUnitService.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Service;
+import se.inera.intyg.infra.integration.hsatk.exception.HsaServiceCallException;
+import se.inera.intyg.infra.integration.intygproxyservice.client.HsaIntygProxyServiceCareProviderOfCareUnitClient;
+import se.inera.intyg.infra.integration.intygproxyservice.dto.GetCareProviderOfCareUnitRequestDTO;
+
+@Service
+@RequiredArgsConstructor
+public class GetCareProviderOfCareUnitService {
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/services/GetCareProviderOfCareUnitService.java:33]: Rename to GetHealthCareUnitService

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/services/GetCareProviderOfCareUnitService.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    private final HsaIntygProxyServiceCareProviderOfCareUnitClient hsaIntygProxyServiceCareProviderOfCareUnitClient;
+
+    @Cacheable(cacheNames = CARE_PROVIDER_OF_CARE_UNIT_CACHE_NAME, key = "#getCareProviderOfCareUnitRequestDTO.careUnitHsaId",
+        unless = "#result == null")
+    public String get(GetCareProviderOfCareUnitRequestDTO getCareProviderOfCareUnitRequestDTO) throws HsaServiceCallException {
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/services/GetCareProviderOfCareUnitService.java:39]: Should return HealthCareUnit.

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
Nice! A few comments to look at though!

**Comment by mikaelaromild1** [comment]
Thanks for the comments! I have made the requested changes :)

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/configuration/CacheConfiguration.java [hunk 11]
```diff
@@ -37,8 +39,24 @@ public class CacheConfiguration {
...
+    @Value("${hsa.intygproxyservice.gethealthunitmembers.cache.expiry:60}")
+    private String hsaHealthUnitMembersCacheExpirySeconds;
+
     @Bean(name = EMPLOYEE_CACHE_NAME)
     public Cache hsaIntygProxyServiceEmployeeCache() {
         return redisCacheOptionsSetter.createCache(EMPLOYEE_CACHE_NAME, hsaEmployeeCacheExpirySeconds);
     }
+
+    @Bean(name = HEALTH_CARE_UNIT_CACHE_NAME)
+    public Cache hsaIntygProxyServiceCareProviderOfCareUnitCache() {
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/configuration/CacheConfiguration.java:54]: Call this hsaIntygProxyServiceHealthCareUnitCache()

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/configuration/CacheConfiguration.java [hunk 12]
```diff
@@ -37,8 +39,24 @@ public class CacheConfiguration {
     @Value("${hsa.intygproxyservice.getemployee.cache.expiry:60}")
     private String hsaEmployeeCacheExpirySeconds;
 
+    @Value("${hsa.intygproxyservice.getcareproviderofcareunit.cache.expiry:60}")
+    private String hsaCareProviderOfCareUnitCacheExpirySeconds;
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/configuration/CacheConfiguration.java:43]: Call this hsaHealtCareUnitCacheExpirySeconds.

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/configuration/CacheConfiguration.java [hunk 13]
```diff
@@ -37,8 +39,24 @@ public class CacheConfiguration {
...
+
+    @Value("${hsa.intygproxyservice.gethealthunitmembers.cache.expiry:60}")
+    private String hsaHealthUnitMembersCacheExpirySeconds;
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/configuration/CacheConfiguration.java:46]: Call this hsaHealtCareUnitMembersCacheExpirySeconds.

File: integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/services/GetHealthCareUnitService.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+import se.inera.intyg.infra.integration.intygproxyservice.client.HsaIntygProxyServiceHealthCareUnitClient;
+import se.inera.intyg.infra.integration.intygproxyservice.dto.GetHealthCareUnitRequestDTO;
+
+@Service
+@RequiredArgsConstructor
+public class GetHealthCareUnitService {
+
+    private final HsaIntygProxyServiceHealthCareUnitClient hsaIntygProxyServiceHealthCareUnitClient;
+
+    @Cacheable(cacheNames = HEALTH_CARE_UNIT_CACHE_NAME, key = "#getCareProviderOfCareUnitRequestDTO.careUnitHsaId",
```
Comment (mww-sopra-steria) [inline integration/hsa-integration-intyg-proxy-service/src/main/java/se/inera/intyg/infra/integration/intygproxyservice/services/GetHealthCareUnitService.java:38]: change #getCareProviderOfCareUnitRequestDTO.careUnitHsaId to #getHealthCareUnitRequestDTO.hsaId

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
Noticed a few more places where naming needs to be updated. Also important that the cache-annotation is correct now when naming has changed.

**Review by mww-sopra-steria — APPROVED** [review]
Approved! :-)

**Pull request merged**